prefect_kubernetes/credentials.py

```diff
@@ -0,0 +1,135 @@
+
+        Returns:
+            client.CoreV1Api: Kubernetes api client to interact with "pod", "service"
+            and "secret" resources
+        """
+        return self.get_kubernetes_client(resource="core")
+
+    def get_batch_client(self) -> client.BatchV1Api:
+        """Convenience method for retrieving a kubernetes api client for job resources
+
+        Returns:
+            client.BatchV1Api: Kubernetes api client to interact with "job" resources
+        """
+        return self.get_kubernetes_client(resource="job")
+
+    def get_app_client(self) -> client.AppsV1Api:
+        """Convenience method for retrieving a kubernetes api client for deployment resources
+
+        Returns:
+            client.AppsV1Api: Kubernetes api client to interact with deployments
+        """
+        return self.get_kubernetes_client(resource="deployment")
+
+    def get_kubernetes_client(self, resource: str) -> KubernetesClient:
+        """
+        Utility function for loading kubernetes client object for a given resource.
+        It will attempt to connect to a Kubernetes cluster in three steps with
+        the first successful connection attempt becoming the mode of communication with
+        a cluster:
+        1. It will first attempt to use a `KubernetesCredentials` block's
+        `cluster_config` to configure a client using
+        `KubernetesClusterConfig.configure_client` and then return the
+        `resource_specific_client`.
+
+        2. Attempt in-cluster connection (will only work when running on a pod)
+        3. Attempt out-of-cluster connection using the default location for a
+        kube config file. In some cases connections to the kubernetes server
+        are dropped after being idle for some time (e.g. Azure Firewall drops
+        idle connections after 4 minutes) which would result in ReadTimeoutErrors.
+
+        In order to prevent that a periodic keep-alive message can be sent to the
+        server to keep the connection open.
+
+        Args:
+            - resource: the name of the resource to retrieve a client for.
+                Currently you can use one of these values: `job`, `pod`, `service`,
+                `deployment`, and `secret`.
+
+        Returns:
+            - KubernetesClient: an initialized, configured Kubernetes Client
+        """
+
+        resource_specific_client = K8S_CLIENTS[resource]
+
+        if self.cluster_config:
+            self.cluster_config.configure_client()
+            return resource_specific_client()
+
+        else:
+            try:
+                print("Trying to load in-cluster configuration...")
+                kube_config.load_incluster_config()
+            except ConfigException as exc:
+                print("{} | Using out of cluster configuration option.".format(exc))
+                print("Loading out-of-cluster configuration...")
+                kube_config.load_kube_config()
```

Comment:
Is it possible to add a test for this snippet?

Reply:
I will eventually add a test here, I'm going to square away support for api key auth first though, because that will affect how I test these